### PR TITLE
Fixes crash when specifying invalid font

### LINKED ISSUE #550: Specifying a font to be used in profile that does not exist causes the app the crash
I added a new profile to the profiles.json specifying a font that is not installed on my system.  When attempting to start the profile the app crashes with the following:

> The program '[15980] WindowsTerminal.exe' has exited with code -1073741819 (0xc0000005) 'Access violation'.

Specifying a font that is installed causes the problem to not occur. 

The profile definition looks like this:

```
    {
      "guid": "{e2f9d7d6-369b-4841-9429-61f5b38474cf}",
      "name": "Bash",
      "colorscheme": "Campbell",
      "historySize": 9001,
      "snapOnInput": true,
      "cursorColor": "#FFFFFF",
      "cursorShape": "bar",
      "commandline": "bash.exe",
      "fontFace": "Fira Code",
      "fontSize": 12,
      "acrylicOpacity": 0.5,
      "useAcrylic": false,
      "closeOnExit": true,
      "padding": "0, 0, 0, 0"
    }

```

### PR DESCRIPTION
**Pull request by miniksa**
## Summary of the Pull Request

Adds fallback routine for when a font name is missing instead of crashing due to uncaught exception thrown when font cannot be found.

## PR Checklist
* [x] Closes #550 
* [x] I'm an employee
* [x] Tests added/passed
* [x] Doesn't need documentation
* [x] I'm a core contributor

## Detailed Description of the Pull Request / Additional comments

For GDI and for DirectX using the stock Layouts and Formats, if your chosen font face name isn't found... something else is just picked on your behalf. When I wrote our custom DX code, I instead threw an exception when I couldn't find the chosen font face. That exception was uncaught and led to a crash.

I decided to just make the custom code act more like the stock code and attempt to choose a fallback if we cannot find what you've requested. I also improved a bit of the passing around of the chosen font name and added locale name information while I was at it.

## Validation Steps Performed
- Opened settings
- Set font name to "arse"
- Watched as nothing changed from previous Consolas selection
- Closed application
- Relaunched
- Didn't crash this time, it just chose Consolas instead

### CODE REVIEW

File: src/renderer/dx/DxRenderer.cpp [hunk 1]
```diff
@@ -1389,11 +1430,107 @@ float DxEngine::GetScaling() const noexcept
...
+    THROW_IF_FAILED(familyNames->GetStringLength(index, &length));
+
+    // https://docs.microsoft.com/en-us/windows/win32/api/dwrite/nf-dwrite-idwritelocalizedstrings-getstringlength
+    // https://docs.microsoft.com/en-us/windows/win32/api/dwrite/nf-dwrite-idwritelocalizedstrings-getstring
+    // Once again, GetStringLength is without the null, but GetString needs the null. So add one.
+    length++;
+
+    // Make our output buffer and resize it so it is allocated.
+    std::wstring retVal;
+    retVal.resize(length);
```
Comment (DHowett-MSFT) [inline src/renderer/dx/DxRenderer.cpp]: do we need to remove the extra terminating null on the wstring after this?

Comment (DHowett-MSFT) [reply src/renderer/dx/DxRenderer.cpp]: er after it is populated

**Comment by DHowett-MSFT** [comment]
Should we cherry-pick this back to 20H1?

File: src/renderer/dx/DxRenderer.cpp [hunk 2]
```diff
@@ -1375,7 +1416,7 @@ float DxEngine::GetScaling() const noexcept
 
     UINT32 familyIndex;
     BOOL familyExists;
-    THROW_IF_FAILED(fontCollection->FindFamilyName(familyName.c_str(), &familyIndex, &familyExists));
+    THROW_IF_FAILED(fontCollection->FindFamilyName(familyName.data(), &familyIndex, &familyExists));
```
[the remark this answers is not in the document]
Comment (miniksa) [reply src/renderer/dx/DxRenderer.cpp]: I think `c_str` is ugly so I change it when I see it.

**Comment by miniksa** [comment]
> Should we cherry-pick this back to 20H1?

Maybe? It's only for the DX renderer which is hidden unless you know the magic dance so it's not high impact. But it is early so it might be an easy sell?

**Comment by miniksa** [comment]
> This seems reasonable to me.
> 
> Should we maybe update SettingsSchem.md with a note that says something like "if we can't find the font specified, we'll try falling back to Consolas"

Sure.

File: src/renderer/dx/DxRenderer.cpp [hunk 3]
```diff
@@ -1354,6 +1356,44 @@ float DxEngine::GetScaling() const noexcept
     return PostMessageW(_hwndTarget, CM_UPDATE_TITLE, 0, (LPARAM) nullptr) ? S_OK : E_FAIL;
 }
 
+// Routine Description:
+// - Attempts to locate the font given, but then begins falling back if we cannot find it.
+// - We'll try to fall back to Consolas with the
```
[the remark this answers is not in the document]
Comment (miniksa) [reply src/renderer/dx/DxRenderer.cpp]: oops lost the train of thought. completed.

**Pull request merged**